chalice/deploy/packager.py

```diff
@@ -208,3 +205,332 @@ def _add_chalice_lib_if_needed(self, project_dir, zip_fileobj):
+        # type: (str, str, Set[Package]) -> None
+        if os.path.isdir(dst_dir):
+            shutil.rmtree(dst_dir)
+        os.makedirs(dst_dir)
+        for whl in whls:
+            zipfile_path = self._osutils.joinpath(src_dir, whl.filename)
+            with zipfile.ZipFile(zipfile_path, 'r') as z:
+                z.extractall(dst_dir)
+
+    def build_site_packages(self, project_dir):
+        # type: (str) -> str
+        requirements_file = self._osutils.joinpath(
+            project_dir, 'requirements.txt')
+        deps_dir = self.site_package_dir(project_dir)
+        if self._has_at_least_one_package(requirements_file):
+            with self._osutils.tempdir() as tempdir:
+                valid_whls, missing_whls = self._download_dependencies(
+                    tempdir, requirements_file)
+                self._install_whls(tempdir, deps_dir, valid_whls)
+            if missing_whls:
+                missing_packages = '\n'.join([p.identifier for p
+                                              in missing_whls])
+                print(MISSING_DEPENDENCIES_TEMPLATE % missing_packages)
+        return deps_dir
+
+    def site_package_dir(self, project_dir):
+        # type: (str) -> str
+        """Return path to the site packages directory."""
+        deps_dir = self._osutils.joinpath(
+            project_dir, '.chalice', 'site-packages')
+        return deps_dir
+
+
+class Package(object):
+    PYPI_SDIST_EXTS = ['.zip', '.tar.gz']
+
+    def __init__(self, directory, filename):
+        # type: (str, str) -> None
+        self.dist_type = 'whl' if filename.endswith('whl') else 'sdist'
+        self.filename = filename
+        self._directory = directory
+        self.identifier = self._calculate_identifier()
+
+    def __str__(self):
+        # type: () -> str
+        return '%s(%s)' % (self.identifier, self.dist_type)
+
+    def __repr__(self):
+        # type: () -> str
+        return str(self)
+
+    def __eq__(self, other):
+        # type: (Any) -> bool
+        return self.identifier == other.identifier
+
+    def __hash__(self):
+        # type: () -> int
+        return hash(self.identifier)
+
+    def _calculate_identifier(self):
+        # type: () -> str
+        # Identiifer that can be fed into pip ie "name==version"
+        if self.dist_type == 'whl':
+            # From the wheel spec (PEP 427)
+            # {distribution}-{version}(-{build tag})?-{python tag}-{abi tag}-
+            # {platform tag}.whl
+            name, version = self.filename.split('-')[:2]
+        else:
+            info_fetcher = SdistMetadataFetcher()
+            name, version = info_fetcher.get_package_name_and_version(
+                self._directory, self.filename)
+        return '%s==%s' % (name, version)
+
+
+class SdistMetadataFetcher(object):
+    """This is the "correct" way to get name and version from an sdist."""
+    # https://github.com/pypa/pip/blob/master/pip/utils/setuptools_build.py
```

Comment:
It may be a reasonable idea to embed a commit hash in this URL since this file can be renamed, deleted, etc in pip's ``master`` branch which would invalidate this link.